### ISSUE TITLE
write a parallel-reading MARC reader

### ISSUE THREAD

**Issue by jrochkind** [traject/traject#188]
@tpendragon reports they did an Elixir MARC indexing that "got to processing 50k records/second without going into Solr (and processing MARC21), and 6k/second actually putting them into solr." 

And one thing that was responsible for this:

> The performance for the Elixir implementation really exploded once we parallelized access of the MARC21 file.

While a ruby indexer is _probably_ (?) not going to be as fast as Elixir, we could use this same technique, and _probably_ get significantly faster.  Perhaps only in JRuby where we don't have the GIL. 

There could be a MARC reader that instead of turning the bytes into MARC serially, reads the bytes from original MARC file serially, but then uses a thread pool to turn each record-bytes-buffer into a ruby-marc object, and then send it into the indexing pipeline.  

This might be tricky and require fibers or something to do as a usual Traject reader, but could be done easily using the new programmatic APIs.  Just have your reader (not really a Traject reader API) do it's thing, and then, in each of it's threads, send the record to the indexer with the new `process_record` method.  The indexer should be thread-safe for use like this. 

For those looking to absolutely maximize performance, this would probably be worth experimenting with.

**Comment by jrochkind** [comment]
We also now support reading more than one file in at once. Right now the files are processed serially, but we could optionally make it process multiple file input parallely, for another approach to parallelizing speed-up when you are able to split your input into multiple files. This might be an even easier approach.

**Comment by billdueber** [comment]
I'm pretty sure I wrote this once. Lemme look and I'll let you know.

**Comment by billdueber** [comment]
OK. A super-simple threaded reader based on the ruby-concurrent Channel routines.

Gist is at https://gist.github.com/billdueber/76b455bbb98c3564aef37e7f2f058911

```
############ TOP LINE RESULTS ########################
# Average of two runs, more or less
#
# mri 2.5    standard  35s
#            threaded  26s (71%)
#
# jruby 9.2  standard  20s
#            threaded  15s (75%)
```

**Comment by jrochkind** [comment]
wait, it's slower? oh no, it's faster but not all that much faster, less than 2x?

**Comment by billdueber** [comment]
Right -- 25-30%. I don't think we'll ever see BEAM level speed up. Our
system threads are just too heavy compared to theirs.

**Comment by jrochkind** [comment]
looking at the code, I don't understand how it's actually doing the parsing in a parallel way. I guess I'll take your word for it? It makes no sense to me that MRI would be faster than JRuby here.

**Comment by billdueber** [comment]
You're reading it wrong -- MRI is 26 seconds threaded, JRuby is 15 seconds. And suddenly I'm not sure I'm doing the parsing in a parallel way -- I think this just might be speedups from having a thread dedicated to sucking the raw records from the filesystem. I'll try again.

**Comment by jrochkind** [comment]
Yeah, that's what I was thinking. 

I'm not sure if there's a sensible way to do the parsing in actual parallel and use the Traject reader API. Might require fibers in a confusing way. You could do a proof of concept without actually using the traject reader API though, something that just parses in parallel and sends to traject with the new `process_record` method.

**Comment by billdueber** [comment]
I tried a couple different ways of threading and a couple different starting conditions (e.g., I loaded the whole file string into memory, and even split on END_OF_RECORD before timing and worked across teh resulting array) and in all cases, threaded decoding was worse. This leads me to believe that the cost of the context switch is too high for the amount of work we're doing during the decode. 

I looked at fibers, but man, I'm not willing to go there. Esp. because I'm selfish and don't actually ever use binary MARC.

The speedup on my original (incorrect) implementation speaks to how slow ruby IO is, but we knew that. 

I'm gonna close this as not worth our time.
